File: tests/integration_e2e.zig
const std = @import("std");
const Context = @import("daigoro").Context;

const exe_path = @import("build_options").exe_path;

comptime {
    if (exe_path == null) {
        @compileError("Integration tests require a path. Run with: zig build integration -Dexe=\"/path/to/ffxiv_dx11.exe\"");
    }
}

const exe = exe_path.?;

test "encode/decode round-trip" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var abs_buf: [std.fs.max_path_bytes]u8 = undefined;
    const cache_dir = try tmp.dir.realpath(".", &abs_buf);

    var ctx = try Context.init(exe, cache_dir);
    defer ctx.deinit();

    const original = "Oodle" ** 200;
    const max_compressed = ctx.maxCompressedSize(original.len) * 2;
    const compressed_buf = try std.testing.allocator.alloc(u8, max_compressed);
    defer std.testing.allocator.free(compressed_buf);

    const written = try ctx.encode(original.ptr, original.len, compressed_buf.ptr, max_compressed);

    const decompressed_buf = try std.testing.allocator.alloc(u8, original.len);
    defer std.testing.allocator.free(decompressed_buf);
    try ctx.decode(compressed_buf.ptr, @intCast(written), decompressed_buf.ptr, original.len);

    try std.testing.expectEqualSlices(u8, original, decompressed_buf);
}

test "rejects double init" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var abs_buf: [std.fs.max_path_bytes]u8 = undefined;
    const cache_dir = try tmp.dir.realpath(".", &abs_buf);

    var ctx = try Context.init(exe, cache_dir);
    defer ctx.deinit();

    try std.testing.expectError(error.AlreadyInitialised, Context.init(exe, cache_dir));
}

test "re-init after destroy" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var abs_buf: [std.fs.max_path_bytes]u8 = undefined;
    const cache_dir = try tmp.dir.realpath(".", &abs_buf);

    var ctx = try Context.init(exe, cache_dir);
    ctx.deinit();

    // Re-init after deinit succeeds
    var ctx2 = try Context.init(exe, cache_dir);
    defer ctx2.deinit();
    try std.testing.expect(ctx2.maxCompressedSize(100) > 0);
}

File: tests/integration_e2e.c
#include "daigoro.h"
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifdef EXE_PATH_MISSING
    #error "Integration tests require a path. Run with: zig build integration -Dexe=\"/path/to/ffxiv_dx11.exe\""
#endif

#ifndef EXE_PATH
    #define EXE_PATH NULL
#endif

int main(void) {
    // Use a uniquely-named temporary cache directory.
    char cache_dir[] = "/tmp/daigoro_e2e_XXXXXX";
    assert(mkdtemp(cache_dir) != NULL);

    // Sanity check
    if (EXE_PATH == NULL) {
        assert(EXE_PATH != NULL && "Missing path: Pass -Dexe to your build command");
    }

    // --- init / encode / decode / destroy ---

    OodleStatus status = oodle_init(EXE_PATH, cache_dir);
    assert(status == OODLE_OK);

    const char src[] = "OodleOodleOodleOodleOodleOodleOodleOodle";
    const uint32_t src_len = (uint32_t)(sizeof(src) - 1);

    const uint32_t max_compressed = oodle_max_compressed_size(src_len);
    uint8_t *compressed = (uint8_t *)malloc(max_compressed);
    assert(compressed != NULL);

    OodleResult r = oodle_encode((const uint8_t *)src, src_len, compressed, max_compressed);
    assert(r.status == OODLE_OK);

    uint8_t *decompressed = (uint8_t *)malloc(src_len);
    assert(decompressed != NULL);

    r = oodle_decode(compressed, r.bytes_written, decompressed, src_len);
    assert(r.status == OODLE_OK);
    assert(memcmp(src, decompressed, src_len) == 0);

    free(decompressed);
    free(compressed);

    // --- double init is rejected ---

    status = oodle_init(EXE_PATH, cache_dir);
    assert(status == OODLE_ERR_ALREADY_INIT);

    // --- destroy and re-init works ---

    oodle_destroy();
    status = oodle_init(EXE_PATH, cache_dir);
    assert(status == OODLE_OK);
    oodle_destroy();

    return 0;
}
